flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/window/combines/RowTimeDeduplicateRecordsCombiner.java

```diff
@@ -0,0 +1,145 @@
+ */
+
+package org.apache.flink.table.runtime.operators.deduplicate.window.combines;
+
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.state.KeyedStateBackend;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.runtime.operators.window.combines.RecordsCombiner;
+import org.apache.flink.table.runtime.operators.window.slicing.WindowTimerService;
+import org.apache.flink.table.runtime.operators.window.state.StateKeyContext;
+import org.apache.flink.table.runtime.operators.window.state.WindowState;
+import org.apache.flink.table.runtime.operators.window.state.WindowValueState;
+import org.apache.flink.table.runtime.util.WindowKey;
+
+import java.util.Iterator;
+
+import static org.apache.flink.table.data.util.RowDataUtil.isAccumulateMsg;
+import static org.apache.flink.table.runtime.operators.deduplicate.DeduplicateFunctionHelper.isDuplicate;
+
+/**
+ * An implementation of {@link RecordsCombiner} that stores the first/last records of incremental
+ * input records into the window state.
+ */
+public final class RowTimeDeduplicateRecordsCombiner implements RecordsCombiner {
+
+    /** The service to register event-time or processing-time timers. */
+    private final WindowTimerService<Long> timerService;
+
+    /** Context to switch current key for states. */
+    private final StateKeyContext keyContext;
+
+    /** The state stores first/last record of each window. */
+    private final WindowValueState<Long> dataState;
+
+    private final int rowtimeIndex;
+
+    private final boolean keepLastRow;
+
+    /** Serializer to copy record if required. */
+    private final TypeSerializer<RowData> recordSerializer;
+
+    public RowTimeDeduplicateRecordsCombiner(
+            WindowTimerService<Long> timerService,
+            StateKeyContext keyContext,
+            WindowValueState<Long> dataState,
+            int rowtimeIndex,
+            boolean keepLastRow,
+            TypeSerializer<RowData> recordSerializer) {
+        this.timerService = timerService;
+        this.keyContext = keyContext;
+        this.dataState = dataState;
+        this.rowtimeIndex = rowtimeIndex;
+        this.keepLastRow = keepLastRow;
+        this.recordSerializer = recordSerializer;
+    }
+
+    @Override
+    public void combine(WindowKey windowKey, Iterator<RowData> records) throws Exception {
+        // step 1: get first/last record of incremental data
+        RowData bufferedResult = null;
+        while (records.hasNext()) {
+            RowData record = records.next();
+            if (!isAccumulateMsg(record)) {
+                throw new UnsupportedOperationException(
+                        "Window deduplicate does not support input RowKind: "
+                                + record.getRowKind().shortString());
+            }
+            if (isDuplicate(bufferedResult, record, rowtimeIndex, keepLastRow)) {
```

Comment:
nit: This `isDuplicate` method name is sort of misleading. It seems like if it returns true we should not update the result. I'd like to propose a name like `shouldUpdate` or such.

Reply:
Maybe it's better to keep the original name because I could not get better name yet. `ShouldUpdate` may misleading we should sent update result to downstream.
Besides, it's a big deal because it's an internal method. WDYT?